runtime/src/main/java/org/corfudb/runtime/object/MVOCache.java

```diff
@@ -0,0 +1,279 @@
+        return objectCache.asMap().containsKey(voId);
+    }
+
+    /**
+     * For a given voId, find in all versions of this object that has the greatest
+     * version && version <= voId.getVersion.
+     *
+     * @param voId the object and version to check
+     * @param snapshotProxyFn snapshotProxy generator
+     * @return an optional of an ICorfuSMRSnapshotProxy instance which is generated
+     *         by the given snapshotProxy generator function. Empty if there is no floorEntry.
+     */
+    public Optional<ICorfuSMRSnapshotProxy<T>> floorEntry(@NonNull VersionedObjectIdentifier voId,
+                                                          @NonNull ISnapshotProxyGenerator<T> snapshotProxyFn) {
+        final TreeSet<Long> allVersionsOfThisObject = objectVersions.get(voId.getObjectId());
+
+        if (allVersionsOfThisObject == null) {
+            // The object has not been created
+            return Optional.empty();
+        } else {
+            synchronized (allVersionsOfThisObject) {
+                final Long floorVersion = allVersionsOfThisObject.floor(voId.getVersion());
+                if (floorVersion == null) {
+                    return Optional.empty();
+                }
+
+                final VersionedObjectIdentifier id = new VersionedObjectIdentifier(voId.getObjectId(), floorVersion);
+                final MVOCacheEntry cacheEntry = objectCache.getIfPresent(id);
+                // Cache eviction updates objectVersions asynchronously.
+                // The item is removed from objectVersions after cache eviction happens.
+                if (cacheEntry == null) {
+                    return Optional.empty();
+                }
+
+                final ICorfuSMRSnapshotProxy<T> snapshotProxy = snapshotProxyFn.generate(id, cacheEntry.getBaseSnapshot());
+                cacheEntry.getSnapshotProxies().add(snapshotProxy);
+                return Optional.of(snapshotProxy);
+            }
+        }
+    }
+
+    /**
+     * Returns true if objectCache has any versionedObject of a certain object.
+     * Checks the objectVersions.
+     *
+     * @param objectId the object id to check
+     * @return true if target object exists
+     */
+    public Boolean containsObject(UUID objectId) {
+        // TODO: what if an object existing in objectCache but its version is not in objectVersions?
+        // This can happen when MVOCache::put is interrupted
+        return objectVersions.containsKey(objectId);
+    }
+
+    public void registerMVO(UUID objectId, MultiVersionObject<T> mvo) {
+        allMVOs.computeIfAbsent(objectId, key -> mvo);
+    }
+
+    private void syncMVOCache() {
+        TokenResponse streamTails = runtime.getSequencerView()
+                .query(allMVOs.keySet().toArray(new UUID[0]));
+        final AtomicInteger count = new AtomicInteger(0);
+        allMVOs.forEach((uuid, mvo) -> {
+            if (objectVersions.get(uuid) != null &&
```

Comment:
This assumes that all the streams can fit in a single message.

Reply:
Why use specific stream tails if you can just query the global tail and sync to that?